core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/BavetConstraintFactory.java

```diff
@@ -42,20 +52,45 @@ public BavetConstraintFactory(SolutionDescriptor<Solution_> solutionDescriptor)
         }
     }
 
+    public <Stream_ extends BavetAbstractConstraintStream<Solution_>> Stream_ share(Stream_ stream) {
+        return share(stream, t -> {});
+    }
+
+    /**
+     * Enables node sharing.
+     * If a constraint already exists in this factory, it replaces it by the old copy.
+     * {@link BavetAbstractConstraintStream} implement equals/hashcode ignoring child streams.
+     * <p>
+     * {@link BavetConstraintSessionFactory#buildSession(boolean, Object)} relies on this occurring for all streams.
+     * <p>
+     * This must be called before the stream receives child streams.
+     * @param stream never null
+     * @param consumer never null
+     * @param <Stream_> the {@link BavetAbstractConstraintStream} subclass
+     * @return  never null
+     */
+    public <Stream_ extends BavetAbstractConstraintStream<Solution_>> Stream_ share(Stream_ stream,
+            Consumer<Stream_> consumer) {
+        return (Stream_) sharingStreamMap.computeIfAbsent(stream, k -> {
+            consumer.accept(stream);
+            return stream;
+        });
+    }
+
     // ************************************************************************
     // from
     // ************************************************************************
 
     @Override
     public <A> UniConstraintStream<A> forEachIncludingNullVars(Class<A> sourceClass) {
         assertValidFromType(sourceClass);
-        return new BavetFromUniConstraintStream<>(this, sourceClass, RetrievalSemantics.STANDARD);
+        return share(new BavetForEachUniConstraintStream<>(this, sourceClass, RetrievalSemantics.STANDARD));
```

Comment:
Note: this `share(...)` is exactly the kind of thing that is easy to forget. In a sense, it is boilerplate. I'm thinking.... is there a way how we can avoid it? A factory class for stream instances could perhaps at least centralize this into one place.

Reply:
It's indeed easy to forgot, but I don't see how to a different approach would make it better, especially because the share() method also accepts a consumer (which is only executed if it's a new stream, not if it already exists). I fear it will only move the problem, not improve it.

I am definitly open for code trying to do that, just don't see how myself. I'll leave this comment open as a reminder.